src/compiler/checker.ts

```diff
@@ -1833,20 +1833,30 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     };
 
     function runWithoutResolvedSignatureCaching<T>(node: Node | undefined, fn: () => T): T {
-        const cachedSignatures = [];
+        const cachedResolvedSignatures = [];
+        const cachedTypes = [];

```

Comment:
I am quite worried that this is not enough to fix all the cases. I feel like focusing on `isCallLikeExpression` and `isFunctionLike` won't handle everything. For example, a reverse mapped type (and its properties) might get cached and a "cleared' function/call might potentially pull from a "corrupted" type. 

I really wonder if this just shouldn't clear **every** cached link on the way from the `node` to the root. It's still hard to determine what exact properties might have to be cleared from links but likely this would cover for more cases and with time the list of cleared (and brought back later) properties would get more complete.

This isn't ideal but with the given architecture, I don't see a better way around this. You had 2 separate checkers in the past but even that was likely prone to similar problems - they were just less apparent (one could request completions, or something else, at 2 different locations within the same call - each of those requests should not depend on the information computed by the other one but it definitely did depend on it in subtle ways).

Reply:
maybe we should just temporarily unset all symbol/node links altogether here? it feels that they should be able to recreate themselves from scratch during the request 🤔

Reply:
After a second though... it's probably not that unlikely that a cached type within the same call~ can mess things up (even if it's like in a separate subtree of the call).

Reply:
> maybe we should just temporarily unset all symbol/node links altogether here? it feels that they should be able to recreate themselves from scratch during the request 🤔

That's more correct, what we have here is a compromise for performance in the common case. Cache invalidation is hard. We want to do 2 things in this function, ideally:

1. Unset any existing, partial result for the signature in question (or just ignore the current result), since we're going to rerun the signature check with some differing flags, and, by extension, any signature whose result depends on it (thus, any syntactic parent, at _least_).
2. Ensure that any calculations we do that are affected by those differing flags (hard to track) aren't cached into the checker long-term. Basically we want to speculate (like we do in the parser), but we lack a real speculation helper inside the checker.

This is... probably _better_, even if it's not a complete fix, just because it invalidates more potentially bad parts of the cache.

Reply:
I had one alternative idea for a potentially more complete fix. Symbol links and node links are primary cache mechanisms related to this class of problems. So maybe we could reassign `getSymbolLinks` and `getNodeLinks` to versions that would not persist results in the "real" links. We could likely still consult the real links for nodes/symbols from different source files. However, we'd use a different "cache" for all nodes/symbols from the current source file. This should avoid broken results altogether as broken results are related to cached information about the nodes close to the one that is the target~ location of the current LSP request.